File: Tree/543. Diameter of Binary Tree/diameterofBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

// diameter:
// distance between any two nodes
// may or may not pass through roots
var diameterOfBinaryTree = function (root) {
    let max = 0;
    dfs(root);
    return max;

    function dfs(root) {
        if (!root) return 0;

        let left = dfs(root.left);
        let right = dfs(root.right);

        // max left (node1) and max right (node2) of the current node
        // diameter
        max = Math.max(max, left + right);
        // height of max left or right subtrees
        return Math.max(left, right) + 1;
    }
};

File: Tree/543. Diameter of Binary Tree/diameterofBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.max = 0
        def dfs(root):
            if not root:
                return 0
            
            left = dfs(root.left)
            right = dfs(root.right)
            
            self.max = max(self.max, left + right)
            return max(left, right) + 1
        
        dfs(root)
        return self.max
        